Cubical/Structures/Group/Morphism.agda

```diff
@@ -0,0 +1,255 @@
+isPropIsGroupHom G H {f} = isPropΠ2 λ a b → Group.is-set H _ _
+
+record GroupHom (G : Group {ℓ}) (H : Group {ℓ'}) : Type (ℓ-max ℓ ℓ') where
+  constructor grouphom
+
+  field
+    fun : ⟨ G ⟩ → ⟨ H ⟩
+    isHom : isGroupHom G H fun
+
+record GroupEquiv (G : Group {ℓ}) (H : Group {ℓ'}) : Type (ℓ-max ℓ ℓ') where
+  constructor groupequiv
+
+  field
+    eq : ⟨ G ⟩ ≃ ⟨ H ⟩
+    isHom : isGroupHom G H (equivFun eq)
+
+  hom : GroupHom G H
+  hom = grouphom (equivFun eq) isHom
+
+-- Morphism composition
+isGroupHomComp : (F : Group {ℓ}) (G : Group {ℓ'}) (H : Group {ℓ''}) →
+  (f : GroupHom F G) → (g : GroupHom G H) → isGroupHom F H (GroupHom.fun g ∘ GroupHom.fun f)
+isGroupHomComp F G H (grouphom f morph-f) (grouphom g morph-g) x y =
+  cong g (morph-f _ _) ∙ morph-g _ _
+
+compGroupHom : {F : Group {ℓ}} {G : Group {ℓ'}} {H : Group {ℓ''}} → GroupHom F G → GroupHom G H → GroupHom F H
+compGroupHom {F = F} {G = G} {H = H} (grouphom f morph-f) (grouphom g morph-g) =
+  grouphom (g ∘ f) (isGroupHomComp F G H (grouphom f morph-f) (grouphom g morph-g))
+
+compGroupEquiv : {F : Group {ℓ}} {G : Group {ℓ'}} {H : Group {ℓ''}} → GroupEquiv F G → GroupEquiv G H → GroupEquiv F H
+compGroupEquiv {F = F} {G = G} {H = H} (groupequiv f morph-f) (groupequiv g morph-g) =
+  groupequiv (compEquiv f g) (isGroupHomComp F G H (grouphom _ morph-f) (grouphom _ morph-g))
+
+idGroupEquiv : (G : Group {ℓ}) → GroupEquiv G G
+idGroupEquiv G = groupequiv (idEquiv (Group.Carrier G)) (λ _ _ → refl)
+
+-- Isomorphism inversion
+isGroupHomInv : (G : Group {ℓ}) (H : Group {ℓ'}) (f : GroupEquiv G H) → isGroupHom H G (invEq (GroupEquiv.eq f))
+isGroupHomInv G H  (groupequiv (f , eq) morph) h h' = isInj-f _ _ (
+  f (g (h ⋆² h') )
+    ≡⟨ retEq (f , eq) _ ⟩
+  h ⋆² h'
+    ≡⟨ sym (λ i → retEq (f , eq) h i ⋆² retEq (f , eq) h' i) ⟩
+  f (g h) ⋆² f (g h')
+    ≡⟨ sym (morph _ _) ⟩
+  f (g h ⋆¹ g h') ∎)
+  where
+  _⋆¹_ = Group._+_ G
+  _⋆²_ = Group._+_ H
+  g = invEq (f , eq)
+  isEmbedding-f : isEmbedding f
+  isEmbedding-f = isEquiv→isEmbedding eq
+  isInj-f : (x y : ⟨ G ⟩) → f x ≡ f y → x ≡ y
+  isInj-f x y = invEq (_ , isEquiv→isEmbedding eq x y)
+
+invGroupEquiv : (G : Group {ℓ}) (H : Group {ℓ'}) → GroupEquiv G H → GroupEquiv H G
+invGroupEquiv G H (groupequiv f morph) = groupequiv (invEquiv f) (isGroupHomInv G H (groupequiv f morph))
+
+groupHomEq : (G : Group {ℓ}) (H : Group {ℓ'}) (f g : GroupHom G H) → (GroupHom.fun f ≡ GroupHom.fun g) → f ≡ g
+groupHomEq G H (grouphom f fm) (grouphom g gm) p i = grouphom (p i) (p-hom i) where
+  p-hom : PathP (λ i → isGroupHom G H (p i)) fm gm
+  p-hom = toPathP (isPropIsGroupHom G H _ _)
+
+groupEquivEq : (G : Group {ℓ}) (H : Group {ℓ'}) (f g : GroupEquiv G H) → (GroupEquiv.eq f ≡ GroupEquiv.eq g) → f ≡ g
+groupEquivEq G H (groupequiv f fm) (groupequiv g gm) p i = groupequiv (p i) (p-hom i) where
+  p-hom : PathP (λ i → isGroupHom G H (p i .fst)) fm gm
+  p-hom = toPathP (isPropIsGroupHom G H _ _)
+
+module GroupΣTheory {ℓ} where
```

Comment:
I would expect this and the rest of the file to be in the Properties file?

Reply:
Oops, I marked this as resolved too quickly. I think this is the last change and then I'll merge

Reply:
I moved it to a separate "MorphismProperties" file for readability.